src/App.tsx

```diff
@@ -1,14 +1,90 @@
 /* eslint-disable max-len */
-import React from 'react';
+import React, {
+  useEffect, useState, useCallback, Dispatch, SetStateAction, useMemo,
+} from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { ErrorMessage } from './components/ErrorMessage';
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [selectedOption, setSelectedOption] = useState<string>('all');
+  const [appliedQuery, setAppliedQuery] = useState<string>('');
+  const [hasRequestError, setHasRequestError] = useState<boolean>(false);
+
+  const loadTodos = async () => {
+    try {
+      const todosFromServer = await getTodos();
+
+      setTodos(todosFromServer);
+      setHasRequestError(false);
+    } catch (error) {
+      setHasRequestError(true);
+    }
+  };
+
+  useEffect(() => {
+    loadTodos();
+  }, []);
+
+  const clearSelectedTodo = () => {
+    setSelectedTodo(null);
+  };
+
+  const selectOption = (option: string) => {
+    setSelectedOption(option);
+  };
+
+  const lowerQuery = appliedQuery.toLowerCase();
+
+  function debounce(f: Dispatch<SetStateAction<string>>, delay: number) {
+    let timerId = 0;
+
+    return (value: string) => {
+      if (timerId) {
+        window.clearTimeout(timerId);
+      }
+
+      timerId = window.setTimeout(() => {
+        f(value);
+      }, delay);
+    };
+  }
+
+  const applyQuery = useCallback(debounce(setAppliedQuery, 800), []);
```

Comment:
api file already have delay

Reply:
Thaaaaanks for such a lot of recommendations! 😊